### PR TITLE
Dis 426 add browser download history for each browser plugin

### PR DESCRIPTION
**Pull request by Wouter-Jansen**

### CODE REVIEW

File: dissect/target/plugins/browsers/firefox.py [hunk 1]
```diff
@@ -114,3 +139,80 @@ def history(self):
...
+            id (string): Record ID.
+            path (string): .
+            url (uri): History URL.
+            size (string): .
+            state (string): .
+            source: (path): The source file of the history record.
+        """
+        for user, db_file, db in self._iter_db("places.sqlite"):
+            try:
+                # TODO: should implement proper indexing in SQLite at some point
```
Comment (pyrco) [inline dissect/target/plugins/browsers/firefox.py:162]: We generally don't leave TODO's in the code, so make it a ticket or fix it ;-)

File: dissect/target/plugins/browsers/firefox.py [hunk 2]
```diff
@@ -114,3 +139,80 @@ def history(self):
...
+                        "type": row.type,
+                        "date_added": from_unix_us(row.dateAdded),
+                        "last_modified": from_unix_us(row.lastModified),
+                    }
+
+                for place_id, annotation in annotations.items():
+                    if "downloads/metaData" not in annotation:
+                        continue
+
+                    place: Row = places[place_id]
```
Comment (pyrco) [inline dissect/target/plugins/browsers/firefox.py:191]:
```suggestion
                    place = places[place_id]
```
We generally don't do type hinting here.

File: dissect/target/plugins/browsers/firefox.py [hunk 3]
```diff
@@ -114,3 +139,80 @@ def history(self):
...
+                    dest_file_info: dict = annotation["downloads/destinationFileURI"]
```
Comment (pyrco) [inline dissect/target/plugins/browsers/firefox.py:192]: Is this guaranteed to be present? Better make it more robust with a `.get()`

File: dissect/target/plugins/browsers/firefox.py [hunk 4]
```diff
@@ -114,3 +139,80 @@ def history(self):
...
+                    metadata: dict = annotation["downloads/metaData"]
+
+                    ended: int = metadata["content"]["endTime"]
```
Comment (pyrco) [inline dissect/target/plugins/browsers/firefox.py:195]: Is this guaranteed to be present? Better make it more robust with a `.get()`

Comment (pyrco) [inline dissect/target/plugins/browsers/firefox.py:195]:
```suggestion
                    ts_end = metadata["content"]["endTime"]
```

File: dissect/target/plugins/browsers/firefox.py [hunk 5]
```diff
@@ -114,3 +139,80 @@ def history(self):
...
+                    ended: datetime = from_unix_ms(ended) if ended else None
+
+                    download_path: str = dest_file_info["content"]
```
Comment (pyrco) [inline dissect/target/plugins/browsers/firefox.py:198]: Is this guaranteed to be present? Better make it more robust with a `.get()`

File: dissect/target/plugins/browsers/firefox.py [hunk 6]
```diff
@@ -114,3 +139,80 @@ def history(self):
...
+
+                    if download_path and self.target.os == "windows":
+                        download_path = path.from_windows(download_path)
+                    elif download_path:
+                        download_path = path(download_path)
+
+                    yield self.BROWSER_DOWNLOAD_RECORD(
+                        ts_start=dest_file_info["date_added"],
```
Comment (pyrco) [inline dissect/target/plugins/browsers/firefox.py:206]: Is this guaranteed to be present? Better make it more robust with a `.get()`

File: dissect/target/plugins/browsers/firefox.py [hunk 7]
```diff
@@ -114,3 +139,80 @@ def history(self):
...
+                        ts_end=ended,
+                        browser="firefox",
+                        id=annotation["id"],
+                        path=download_path,
+                        url=try_idna(place.url),
+                        size=metadata["content"]["fileSize"],
```
Comment (pyrco) [inline dissect/target/plugins/browsers/firefox.py:212]: Is this guaranteed to be present? Better make it more robust with a `.get()`

File: dissect/target/plugins/browsers/firefox.py [hunk 8]
```diff
@@ -114,3 +139,80 @@ def history(self):
...
+                        state=metadata["content"]["state"],
```
Comment (pyrco) [inline dissect/target/plugins/browsers/firefox.py:213]: Is this guaranteed to be present? Better make it more robust with a `.get()`

**Review by pyrco — CHANGES_REQUESTED** [review]

File: dissect/target/plugins/browsers/chrome.py [hunk 9]
```diff
@@ -23,8 +26,16 @@ class ChromePlugin(ChromiumMixin, Plugin):
...
 
     @export(record=HISTORY_RECORD)
     def history(self):
         """Return browser history records for Google Chrome."""
         yield from ChromiumMixin.history(self, "chrome")
+
+    @export(record=BROWSER_DOWNLOAD_RECORD)
+    def downloads(self):
+        """Return browser download records for Google Chrome."""
+        yield from ChromiumMixin.downloads(self, "chrome")
```
[the remark this answers is not in the document]
Comment (pyrco) [reply dissect/target/plugins/browsers/chrome.py:41]: Ehm I mean `super().downloads("chrome") of course.

File: dissect/target/plugins/browsers/chromium.py [hunk 10]
```diff
@@ -155,3 +214,8 @@ class ChromiumPlugin(ChromiumMixin, Plugin):
     def history(self):
         """Return browser history records for Chromium browser."""
         yield from ChromiumMixin.history(self, "chromium")
+
+    @export(record=ChromiumMixin.BROWSER_DOWNLOAD_RECORD)
+    def downloads(self):
+        """Return browser download records for Chromium browser."""
+        yield from ChromiumMixin.downloads(self, "chromium")
```
[the remark this answers is not in the document]
Comment (pyrco) [reply dissect/target/plugins/browsers/chromium.py:221]: Ehm I mean `super().downloads("chromium")` of course.

File: dissect/target/plugins/browsers/edge.py [hunk 11]
```diff
@@ -19,8 +22,16 @@ class EdgePlugin(ChromiumMixin, Plugin):
...
 
     @export(record=HISTORY_RECORD)
     def history(self):
         """Return browser history records for Microsoft Edge."""
         yield from ChromiumMixin.history(self, "edge")
+
+    @export(record=BROWSER_DOWNLOAD_RECORD)
+    def downloads(self):
+        """Return browser download records for Microsoft Edge."""
+        yield from ChromiumMixin.downloads(self, "edge")
```
[the remark this answers is not in the document]
Comment (pyrco) [reply dissect/target/plugins/browsers/edge.py:37]: Ehm I mean `super().downloads("edge")` of course.

**Review by pyrco — CHANGES_REQUESTED** [review]

**Comment by pyrco** [comment]
The binary files used for tests should be made in to git-lfs files.

File: .gitattributes [hunk 12]
```diff
@@ -1,2 +1,7 @@
 tests/data/enc-volume.bin filter=lfs diff=lfs merge=lfs -text
 tests/data/uppercase_driveletter.tar filter=lfs diff=lfs merge=lfs -text
+tests/data/WebCacheV01.dat.gz filter=lfs diff=lfs merge=lfs -text
```
Comment (Schamper) [inline .gitattributes:3]: Can you move these files to:
```
tests/data/plugins/browsers/iexplore/WebCacheV01.dat.gz
tests/data/plugins/browsers/firefox/places.sqlite
tests/data/plugins/browsers/chrome/History.sqlite
tests/data/plugins/browsers/edge/History.sqlite
tests/data/plugins/browsers/chromium/History.sqlite
```
?

**Review by Schamper — DISMISSED** [review]

**Comment by pyrco** [comment]
LGTM

**Review by pyrco — APPROVED** [review]

**Pull request merged**